File: cpp/tests/test_plat_stats.cpp
/// PLAT and native-statistics tests: rank/Spearman/KS hand cases and tie
/// handling, pinned zone thresholds at the exact boundaries, the constant-
/// series Red convention, and the Amber surcharge formula.

#include <gtest/gtest.h>

#include <cmath>

#include "frtb/plat.hpp"
#include "frtb/stats.hpp"
#include "test_helpers.hpp"

namespace {

using namespace frtb;

TEST(Stats, AverageRanksWithTies) {
    // Values 10, 20, 20, 30 -> ranks 1, 2.5, 2.5, 4.
    std::vector<double> r = average_ranks({10.0, 20.0, 20.0, 30.0});
    EXPECT_DOUBLE_EQ(r[0], 1.0);
    EXPECT_DOUBLE_EQ(r[1], 2.5);
    EXPECT_DOUBLE_EQ(r[2], 2.5);
    EXPECT_DOUBLE_EQ(r[3], 4.0);
}

TEST(Stats, SpearmanMonotone) {
    std::vector<double> x = {1.0, 2.0, 3.0, 4.0, 5.0};
    std::vector<double> y_up = {10.0, 20.0, 25.0, 40.0, 100.0};
    std::vector<double> y_dn = {5.0, 4.0, 3.0, 2.0, 1.0};
    EXPECT_NEAR(spearman(x, y_up), 1.0, 1e-12);
    EXPECT_NEAR(spearman(x, y_dn), -1.0, 1e-12);
}

TEST(Stats, PearsonErrors) {
    EXPECT_THROW(pearson({1.0, 1.0, 1.0}, {1.0, 2.0, 3.0}), std::invalid_argument);  // constant
    EXPECT_THROW(pearson({1.0}, {2.0}), std::invalid_argument);                      // too short
    EXPECT_THROW(pearson({1.0, 2.0}, {1.0, 2.0, 3.0}), std::invalid_argument);       // mismatch
    EXPECT_THROW(spearman({1.0, NAN, 3.0}, {1.0, 2.0, 3.0}), std::invalid_argument);
}

TEST(Stats, KsHandCases) {
    // Identical samples: D = 0.  Fully separated samples: D = 1.
    std::vector<double> a = {1.0, 2.0, 3.0, 4.0};
    EXPECT_DOUBLE_EQ(ks_statistic(a, a), 0.0);
    EXPECT_DOUBLE_EQ(ks_statistic({1.0, 2.0}, {10.0, 20.0}), 1.0);
    // Hand case with ties across samples: max pooled ECDF gap = 0.25.
    EXPECT_DOUBLE_EQ(ks_statistic({1.0, 2.0, 3.0, 4.0}, {2.0, 3.0, 4.0, 5.0}), 0.25);
    EXPECT_THROW(ks_statistic({}, {1.0}), std::invalid_argument);
}

TEST(Plat, ZoneThresholdBoundaries) {
    const SbmParams& p = frtb_test::params();
    // Exact green boundary: spearman = 0.85, KS = 0.09.
    EXPECT_EQ(plat_zone_from_metrics(0.85, 0.09, p), "green");
    // Just below green spearman / above green KS -> amber.
    EXPECT_EQ(plat_zone_from_metrics(0.8499, 0.09, p), "amber");
    EXPECT_EQ(plat_zone_from_metrics(0.85, 0.0901, p), "amber");
    // Amber floor is inclusive: spearman = 0.80, KS = 0.12 stays amber.
    EXPECT_EQ(plat_zone_from_metrics(0.80, 0.12, p), "amber");
    // Below the amber floor / above the amber KS cap -> red.
    EXPECT_EQ(plat_zone_from_metrics(0.7999, 0.05, p), "red");
    EXPECT_EQ(plat_zone_from_metrics(0.9, 0.1201, p), "red");
    EXPECT_THROW(plat_zone_from_metrics(NAN, 0.05, p), std::invalid_argument);
}

TEST(Plat, ConstantSeriesIsRedWithNullMetrics) {
    const SbmParams& p = frtb_test::params();
    std::vector<double> constant(10, 3.0);
    std::vector<double> varying = {1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0};
    PlatResult r = plat_test(constant, varying, p);
    EXPECT_EQ(r.zone, "red");
    EXPECT_FALSE(r.spearman.has_value());
    EXPECT_FALSE(r.ks.has_value());
    EXPECT_THROW(plat_test({1.0, 2.0}, {1.0, 2.0}, p), std::invalid_argument);  // < 3 obs
    EXPECT_THROW(plat_test({1.0, 2.0, 3.0}, {1.0, 2.0}, p), std::invalid_argument);
}

TEST(Plat, SurchargeFormula) {
    const SbmParams& p = frtb_test::params();
    EXPECT_DOUBLE_EQ(plat_surcharge("amber", 300.0, 100.0, p), 100.0);  // 0.5*(300-100)
    EXPECT_DOUBLE_EQ(plat_surcharge("amber", 100.0, 300.0, p), 0.0);    // floored at 0
    EXPECT_DOUBLE_EQ(plat_surcharge("green", 300.0, 100.0, p), 0.0);
    EXPECT_DOUBLE_EQ(plat_surcharge("red", 300.0, 100.0, p), 0.0);
    EXPECT_THROW(plat_surcharge("blue", 1.0, 1.0, p), std::invalid_argument);
    EXPECT_THROW(plat_surcharge("amber", -1.0, 1.0, p), std::invalid_argument);
}

TEST(Plat, BundledDesksMatchPinnedFacts) {
    // desk1 green (no surcharge), desk2 amber (positive surcharge since SA > IMA core).
    const frtb::Results& res = frtb_test::results();
    EXPECT_EQ(res.ima.at("desk1").plat.zone, "green");
    EXPECT_EQ(res.ima.at("desk2").plat.zone, "amber");
    EXPECT_DOUBLE_EQ(res.ima.at("desk1").plat_surcharge, 0.0);
    EXPECT_GT(res.ima.at("desk2").plat_surcharge, 0.0);
}

}  // namespace

File: python/tests/test_plat_stats.py
"""PLAT zones/surcharge and the native Spearman/KS implementations
(cross-checked against scipy — scipy is used ONLY here, in tests)."""
import math

import pytest
from scipy import stats as sps

from frtb.plat import plat_surcharge, plat_test, plat_zone_from_metrics
from frtb.stats import average_ranks, ks_statistic, pearson, spearman


def series(n: int, f1: float, f2: float, scale: float = 1.0):
    """Deterministic pseudo-noisy series (no RNG in the test suite)."""
    return [scale * (math.sin(f1 * i) + 0.5 * math.cos(f2 * i * i)) for i in range(n)]


class TestNativeStats:
    def test_spearman_vs_scipy_grid(self):
        # property-style loop over several deterministic series pairs
        for n in (25, 60, 120):
            for (a1, a2), (b1, b2) in [((1.1, 0.3), (0.9, 0.7)),
                                       ((2.3, 0.11), (2.3, 0.13)),
                                       ((0.5, 1.7), (3.1, 0.2))]:
                x = series(n, a1, a2)
                y = [u + v for u, v in zip(series(n, b1, b2), x)]
                want = sps.spearmanr(x, y).statistic
                assert spearman(x, y) == pytest.approx(want, abs=1e-12)

    def test_spearman_with_ties_vs_scipy(self):
        x = [1.0, 2.0, 2.0, 3.0, 3.0, 3.0, 4.0, 0.0]
        y = [5.0, 5.0, 6.0, 7.0, 8.0, 8.0, 9.0, 4.0]
        assert spearman(x, y) == pytest.approx(sps.spearmanr(x, y).statistic, abs=1e-12)

    def test_average_ranks_ties(self):
        assert average_ranks([10.0, 20.0, 20.0, 30.0]) == [1.0, 2.5, 2.5, 4.0]

    def test_perfect_monotone(self):
        x = [1.0, 2.0, 3.0, 4.0]
        assert spearman(x, [10.0, 20.0, 30.0, 40.0]) == pytest.approx(1.0, abs=1e-15)
        assert spearman(x, [4.0, 3.0, 2.0, 1.0]) == pytest.approx(-1.0, abs=1e-15)

    def test_ks_vs_scipy_grid(self):
        for n, m in ((30, 30), (50, 80), (120, 40)):
            x = series(n, 1.3, 0.21)
            y = [v + 0.3 for v in series(m, 0.7, 0.4)]
            want = sps.ks_2samp(x, y).statistic
            assert ks_statistic(x, y) == pytest.approx(want, abs=1e-12)

    def test_ks_with_ties_vs_scipy(self):
        x = [1.0, 1.0, 2.0, 3.0]
        y = [1.0, 2.0, 2.0, 4.0]
        assert ks_statistic(x, y) == pytest.approx(sps.ks_2samp(x, y).statistic, abs=1e-12)

    def test_ks_identical_zero(self):
        x = series(40, 1.0, 0.5)
        assert ks_statistic(x, list(x)) == 0.0

    def test_errors(self):
        with pytest.raises(ValueError, match="equal length"):
            spearman([1.0, 2.0, 3.0], [1.0, 2.0])
        with pytest.raises(ValueError, match="constant"):
            pearson([1.0, 1.0, 1.0], [1.0, 2.0, 3.0])
        with pytest.raises(ValueError, match="non-empty"):
            ks_statistic([], [1.0])
        with pytest.raises(ValueError, match="finite"):
            ks_statistic([float("nan")], [1.0])


class TestPlatZones:
    def test_threshold_boundaries(self, params):
        z = lambda sp, ks: plat_zone_from_metrics(sp, ks, params)
        assert z(0.85, 0.09) == "green"          # exactly on both green edges
        assert z(0.8499999, 0.09) == "amber"     # spearman just below green
        assert z(0.85, 0.0900001) == "amber"     # ks just above green
        assert z(0.80, 0.12) == "amber"          # exactly on both amber edges
        assert z(0.7999999, 0.05) == "red"       # spearman below amber
        assert z(0.99, 0.1200001) == "red"       # ks above amber
        assert z(0.82, 0.05) == "amber"

    def test_constant_pnl_is_red(self, params):
        res = plat_test([5.0] * 20, series(20, 1.1, 0.3), params)
        assert res.zone == "red"
        assert res.spearman is None and res.ks is None
        # constant on the RTPL side too
        assert plat_test(series(20, 1.1, 0.3), [0.0] * 20, params).zone == "red"

    def test_identical_series_green(self, params):
        x = series(50, 1.3, 0.7)
        res = plat_test(x, list(x), params)
        assert res.zone == "green"
        assert res.spearman == pytest.approx(1.0, abs=1e-12)
        assert res.ks == 0.0

    def test_length_errors(self, params):
        with pytest.raises(ValueError, match="mismatch"):
            plat_test([1.0, 2.0], [1.0], params)
        with pytest.raises(ValueError, match="at least 3"):
            plat_test([1.0, 2.0], [1.0, 2.0], params)


class TestPlatSurcharge:
    def test_amber_interpolation(self, params):
        # k = 0.5 pinned: surcharge = 0.5 * max(0, SA - IMA)
        assert plat_surcharge("amber", 100.0, 60.0, params) == pytest.approx(20.0, abs=1e-12)
        assert plat_surcharge("amber", 50.0, 60.0, params) == 0.0  # SA < IMA

    def test_green_red_no_surcharge(self, params):
        assert plat_surcharge("green", 100.0, 60.0, params) == 0.0
        assert plat_surcharge("red", 100.0, 60.0, params) == 0.0

    def test_bad_zone_raises(self, params):
        with pytest.raises(ValueError, match="zone"):
            plat_surcharge("blue", 1.0, 1.0, params)
